api/graphql/mutations/user.js

```diff
@@ -1,84 +1,194 @@
+    let user = await User.find(email, {}, false)
+
+    // Email already validated: 
+    // send failure without error to not reveal account existence
+    if (user?.get('email_validated')) {
+      return { success: false }
+    }
+
+    // User is new or exists without a validated email:
+    // send verification email
+    if (!user) {
+      user = await User.create({ email, active: false })
+    }
+
+    // TODO: Check here if a non-expired UserVerificationCode already exists for this user
+    //       if so extend expiration for another 4 hours and resend that code?
+    const { code, token } = await UserVerificationCode.create(email)
+
+    Queue.classMethod('Email', 'sendEmailVerification', {
+      email,
+      version: 'with link',
+      templateData: {
+        code,
+        verify_url: Frontend.Route.verifyEmail(email, token)
       }
+    })
+
+    return { success: true }
+  } catch (error) {
+    return { success: false }
+  }
+}
+
+export const verifyEmail = (fetchOne, { req }) => async (_, { code, email, token }) => {
+  try {
+    const codeVerified = await UserVerificationCode.verify({
+      email,
+      code,
+      token
+    })
+
+    if (!codeVerified) {
+      return { error: 'Invalid code, please try again' }
     }
-    const user = await User.authenticate(email, password)
-    await UserSession.login(req, user, 'password')
+
+    const user = await User.find(email, {}, false)
+
+    await user.save({ email_validated: true })
+
+    req.session.userId = user.id
+
     return {
       me: fetchOne('Me', user.id)
     }
-  } catch(err) {
-    throw new Error(err.message)
+  } catch (error) {
+    return { error: 'Link expired, please start over' }
   }
 }
 
-export async function register({ req }, { name, password }) {
-  let user = await User.find(req.session.userId, {}, false)
-  if (!user) {
-    throw new Error("Not authorized")
-  }
-  if (!user.get('email_validated')) {
-    throw new Error("Email not validated")
-  }
+export const register = (fetchOne, { req }) => async (_, { name, password }) => {
+  try {
+    const user = await User.find(req.session.userId, {}, false)
 
-  return bookshelf.transaction(transacting =>
-    user.save({ name, active: true }, { transacting }).then(async (user) => {
+    if (!user) {
+      return { error: 'Not authorized' }
+    }
+
+    if (!user.get('email_validated')) {
+      return { error: 'Email not validated' }
+    }
+
+    await bookshelf.transaction(async transacting => {
+      await user.save({ name, active: true }, { transacting })
       await UserSession.login(req, user, 'password', { transacting }) // XXX: this does another save of the user, ideally we just do one of those
       await LinkedAccount.create(user.id, { type: 'password', password }, { transacting })
       await Analytics.trackSignup(user.id, req)
-      return user
-    })
-    .catch(function (err) {
-      console.log("Error registering user", err.message)
-      throw err
```

Comment:
i see you are switching from throwing errors to returning { error: message }, does this work the same? if you throw an error it returns an error status code whereas just returning JSON will not

Reply:
No, it will need to be and is going to be handled differently in the front-end. Currently finishing-up both Web and Mobile work to accommodate these updates. Throwing errors in a graphql resolver creates a spec defined errors array which is really handy, but in the last couple years this has fallen out of favor as the best way to handle errors. I don't have a succinct summary of the reasons why, but I was convinced enough when I had to make the call one way or another to prefer handling errors this way. Here are a couple articles on the topic, both will take it further than what we're doing, but give good explanations of the downsides of relying on the graphql default error object: https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces , https://productionreadygraphql.com/2020-08-01-guide-to-graphql-errors ... 🤷🏼‍♂️ I prefer this for now as it makes handling the errors at the component level a little simpler vs retrieving them out of redux after the reducer stores them.

Reply:
We will likely want to convert our current `error` types to `errors` as an array of `error`s and make each error have a code and message option. What is here is along the way, if that makes sense.